cmake-tool/helpers/cpio.cmake

```diff
@@ -39,48 +39,100 @@ function(MakeCPIO output_name input_files)
+    # Since some application access the archive contents by index and not by
+    # name, the files must be put into the archive in exactly the same order as
+    # they are listen in 'input_files'. Using simply "ls | cpio ...." instead of
+    # "printf '%s\\n' \${@##*/} | cpio ..." does not guarantee preserving the
+    # order.
     CheckCPIOArgument(cpio_reproducible_flag "--reproducible")
-    set(
-        commands
-        "bash;-c;cpio ${cpio_reproducible_flag} --quiet --create -H newc --file=${CMAKE_CURRENT_BINARY_DIR}/archive.${output_name}.cpio;&&"
+    set(cpio_archive_creator "${output_name_abs}.sh")
+    file(
+        WRITE
+            "${cpio_archive_creator}"
+            # content
+            "#!/bin/sh\n"
+            "# auto-generated file from MakeCPIO(), do not edit\n"
+            "TMP_DIR=${cpio_archive}.files\n"
+            "mkdir -p \${TMP_DIR}\n"
+            "cp -a \"$@\" \${TMP_DIR}/\n"
+            "touch -d 1970-01-01T00:00:00Z \${TMP_DIR}/*\n"
+            "(\n"
+            "  cd \${TMP_DIR}\n"
+            "  printf '%s\\n' \${@##*/} | cpio ${cpio_reproducible_flag} --quiet --create --format=newc --owner=+0:+0\n"
+            ") > ${cpio_archive}\n"
+            "rm -rf \${TMP_DIR}\n"
     )
-    foreach(file IN LISTS input_files)
-        # Try and generate reproducible cpio meta-data as we do this:
-        # - touch -d @0 file sets the modified time to 0
-        # - --owner=root:root sets user and group values to 0:0
-        # - --reproducible creates reproducible archives with consistent inodes and device numbering
-        list(
-            APPEND
-                commands
-                "bash;-c; mkdir -p temp_${output_name} && cd temp_${output_name} && cp -a ${file} . && touch -d 1970-01-01T00:00:00Z `basename ${file}` && echo `basename ${file}` | cpio --append ${cpio_reproducible_flag} --owner=+0:+0 --quiet -o -H newc --file=${CMAKE_CURRENT_BINARY_DIR}/archive.${output_name}.cpio && rm `basename ${file}` && cd ../ && rmdir temp_${output_name};&&"
-        )
-    endforeach()
-    list(APPEND commands "true")
+
+    # Create a "program" that makes the compiler generate and object file that
+    # contains the cpio archive.
+    # CMake wants the absolute name when creating files
+    set(cpio_archive_s "${output_name_abs}.S")
+    file(
+        WRITE
+            "${cpio_archive_s}"
+            # content
+            "# auto-generated file from MakeCPIO(), do not edit\n"
+            ".section ._archive_cpio, \"aw\"\n"
+            ".globl ${archive_symbol}, ${archive_symbol}_end\n"
+            "${archive_symbol}:\n"
+            ".incbin \"${cpio_archive}\"\n"
+            "${archive_symbol}_end:\n"
+    )
+
+    # Re-run CMake configuration in case 'cpio_archive_creator' or
+    # 'cpio_archive_s' is deleted.
+    set_property(
+        DIRECTORY
+        APPEND
+        PROPERTY CMAKE_CONFIGURE_DEPENDS "${cpio_archive_creator}" "${cpio_archive_s}"
+    )
```

Comment:
I don't consider that this property is needed unless CMake implicitly adds it for all files it generates in the build directory already.  Trying to detect and recover the build directory becoming corrupted isn't the responsibility of the build scripts.

Reply:
It's a convenience feature I found quite useful when experimenting with the archive creation. In general, when debugging CMake things, I often corrupt the build folder or delete files on purpose to force partial regeneration. And my expectation from proper dependency management is, that everything in the build folder should be re-generated automatically.